Compile failure on MacOSX

### What component of `google-cloud-cpp` is this related to?

Remove the ones that do not apply.

- Cloud Storage.

### What version of `google-cloud-cpp` are you using?

b7af0a917f84cfd5401869b40864699648021a30

### What compiler and version are you using?

$ c++ -v
Apple LLVM version 10.0.1 (clang-1001.0.46.4)
Target: x86_64-apple-darwin18.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
### What operating system and version are you using?

MacOSX 10.14.5

### What were you trying to do?

Compile fails with the following error

```
/Users/tmatsuo/work/google-cloud-cpp/google/cloud/storage/benchmarks/benchmark_utils.cc:207:19: error: use of undeclared identifier 'RUSAGE_THREAD'
  (void)getrusage(RUSAGE_THREAD, &start_usage_);
                  ^
/Users/tmatsuo/work/google-cloud-cpp/google/cloud/storage/benchmarks/benchmark_utils.cc:222:19: error: use of undeclared identifier 'RUSAGE_THREAD'
  (void)getrusage(RUSAGE_THREAD, &now);
                  ^
2 errors generated.
```

### What did you expect to see?

Compile succeeds on MacOSX. Actually it successfully compiles when passing `-DGOOGLE_CLOUD_CPP_HAS_GETRUSAGE=0`, but it is nicer if it compiles without it.

`RUSAGE_THREAD` seems like linux only flag.

--- a/google/cloud/storage/benchmarks/CMakeLists.txt
+++ b/google/cloud/storage/benchmarks/CMakeLists.txt
@@ -21,10 +21,15 @@ target_link_libraries(storage_benchmarks
                               google_cloud_cpp_common_options)
 
 include(CheckSymbolExists)
-check_symbol_exists(getrusage sys/resource.h GOOGLE_CLOUD_CPP_HAS_GETRUSAGE)
+check_symbol_exists(getrusage sys/resource.h GOOGLE_CLOUD_CPP_HAVE_GETRUSAGE)
+check_symbol_exists(RUSAGE_THREAD sys/resource.h
+                    GOOGLE_CLOUD_CPP_HAVE_RUSAGE_THREAD)
 target_compile_definitions(
-    storage_benchmarks PUBLIC
-    GOOGLE_CLOUD_CPP_HAS_GETRUSAGE=$<BOOL:${GOOGLE_CLOUD_CPP_HAS_GETRUSAGE}>)
+    storage_benchmarks
+    PUBLIC
+    GOOGLE_CLOUD_CPP_HAVE_GETRUSAGE=$<BOOL:${GOOGLE_CLOUD_CPP_HAVE_GETRUSAGE}>
+    GOOGLE_CLOUD_CPP_HAVE_RUSAGE_THREAD=$<BOOL:${GOOGLE_CLOUD_CPP_HAVE_RUSAGE_THREAD}>
+    )
 
 include(CreateBazelConfig)
 create_bazel_config(storage_benchmarks YEAR 2019)

--- a/google/cloud/storage/benchmarks/benchmark_utils.cc
+++ b/google/cloud/storage/benchmarks/benchmark_utils.cc
@@ -194,9 +194,19 @@ std::vector<std::string> OptionsParse(std::vector<OptionDescriptor> const& desc,
   return argv;
 }
 
+namespace {
+int rusage_who() {
+#if GOOGLE_CLOUD_CPP_HAVE_RUSAGE_THREAD
+  return RUSAGE_THREAD;
+#else
+  return RUSAGE_SELF;
+#endif  // GOOGLE_CLOUD_CPP_HAVE_RUSAGE_THREAD
+}
+}  // namespace
+
 void SimpleTimer::Start() {
-#if GOOGLE_CLOUD_CPP_HAS_GETRUSAGE
-  (void)getrusage(RUSAGE_THREAD, &start_usage_);
+#if GOOGLE_CLOUD_CPP_HAVE_GETRUSAGE
+  (void)getrusage(rusage_who(), &start_usage_);
 #endif  // GOOGLE_CLOUD_CPP_HAVE_GETRUSAGE
   start_ = std::chrono::steady_clock::now();
 }
@@ -205,13 +215,13 @@ void SimpleTimer::Stop() {
   using namespace std::chrono;
   elapsed_time_ = duration_cast<microseconds>(steady_clock::now() - start_);
 
-#if GOOGLE_CLOUD_CPP_HAS_GETRUSAGE
+#if GOOGLE_CLOUD_CPP_HAVE_GETRUSAGE
   auto as_usec = [](timeval const& tv) {
     return microseconds(seconds(tv.tv_sec)) + microseconds(tv.tv_usec);
   };
 
   struct rusage now {};
-  (void)getrusage(RUSAGE_THREAD, &now);
+  (void)getrusage(rusage_who(), &now);
   auto utime = as_usec(now.ru_utime) - as_usec(start_usage_.ru_utime);
   auto stime = as_usec(now.ru_stime) - as_usec(start_usage_.ru_stime);
   cpu_time_ = utime + stime;
@@ -250,7 +260,15 @@ void SimpleTimer::Stop() {
      << "# voluntary context switches   =" << now.ru_nvcsw << "\n"
      << "# involuntary context switches =" << now.ru_nivcsw << "\n";
   annotations_ = std::move(os).str();
-#endif  // GOOGLE_CLOUD_CPP_HAS_GETRUSAGE
+#endif  // GOOGLE_CLOUD_CPP_HAVE_GETRUSAGE
+}
+
+bool SimpleTimer::SupportPerThreadUsage() {
+#if GOOGLE_CLOUD_CPP_HAVE_RUSAGE_THREAD
+  return true;
+#else
+  return false;
+#endif  // GOOGLE_CLOUD_CPP_HAVE_RUSAGE_THREAD
 }
 
 }  // namespace storage_benchmarks

--- a/google/cloud/storage/benchmarks/benchmark_utils.cc
+++ b/google/cloud/storage/benchmarks/benchmark_utils.cc
@@ -194,6 +194,7 @@ std::vector<std::string> OptionsParse(std::vector<OptionDescriptor> const& desc,
   return argv;
 }
 
+#if GOOGLE_CLOUD_CPP_HAVE_GETRUSAGE
 namespace {
 int rusage_who() {
 #if GOOGLE_CLOUD_CPP_HAVE_RUSAGE_THREAD
@@ -203,6 +204,7 @@ int rusage_who() {
 #endif  // GOOGLE_CLOUD_CPP_HAVE_RUSAGE_THREAD
 }
 }  // namespace
+#endif  // GOOGLE_CLOUD_CPP_HAVE_GETRUSAGE
 
 void SimpleTimer::Start() {
 #if GOOGLE_CLOUD_CPP_HAVE_GETRUSAGE

--- a/google/cloud/storage/benchmarks/benchmark_utils.h
+++ b/google/cloud/storage/benchmarks/benchmark_utils.h
@@ -19,9 +19,9 @@
 #include <chrono>
 #include <functional>
 #include <string>
-#if GOOGLE_CLOUD_CPP_HAS_GETRUSAGE
+#if GOOGLE_CLOUD_CPP_HAVE_GETRUSAGE
 #include <sys/resource.h>
-#endif  // GOOGLE_CLOUD_CPP_HAS_GETRUSAGE
+#endif  // GOOGLE_CLOUD_CPP_HAVE_GETRUSAGE
 
 namespace google {
 namespace cloud {
@@ -103,13 +103,15 @@ class SimpleTimer {
   std::string const& annotations() const { return annotations_; }
   //@}
 
+  static bool SupportPerThreadUsage();
+
  private:
   std::chrono::steady_clock::time_point start_;
   std::chrono::microseconds elapsed_time_;
   std::chrono::microseconds cpu_time_;
-#if GOOGLE_CLOUD_CPP_HAS_GETRUSAGE
+#if GOOGLE_CLOUD_CPP_HAVE_GETRUSAGE
   struct rusage start_usage_;
-#endif  // GOOGLE_CLOUD_CPP_HAS_GETRUSAGE
+#endif  // GOOGLE_CLOUD_CPP_HAVE_GETRUSAGE
   std::string annotations_;
 };
 

--- a/google/cloud/storage/benchmarks/storage_throughput_vs_cpu_benchmark.cc
+++ b/google/cloud/storage/benchmarks/storage_throughput_vs_cpu_benchmark.cc
@@ -407,6 +407,14 @@ Options ParseArgs(int argc, char* argv[]) {
     throw std::runtime_error(std::move(os).str());
   }
 
+  if (!gcs_bm::SimpleTimer::SupportPerThreadUsage() &&
+      options.thread_count > 1) {
+    std::ostringstream os;
+    os << "Your platform does not support per-thread usage metrics"
+          " (see getrusage(2)). Running more than one thread is not supported.";
+    throw std::runtime_error(std::move(os).str());
+  }
+
   return options;
 }
 

